pinctrl/pinctrl_utils.go

```diff
@@ -279,82 +303,90 @@ func (b *pinctrlpi5) createGPIOVPage(memPath string) error {
+	*/
 
-		In this implementation, we know that our base address is aligned with the beginning of a page. However,
-		below is the implementation required when they do not align.
-
+	var offset int64 // default 0 offset
+	lenMapping := int(chipSize)
+	// adjust the map parameters to handle the misalignment
+	if !useGPIOMem {
 		pageSize := uint64(syscall.Getpagesize())
-		pageStart := b.physAddr & (pageSize - 1)
+		pageStart := physAddr & (pageSize - 1)
 		// difference between base address of the page and the address we're actually tring to access
-		dataStartingAddrDiff = b.physAddr - pageStart
-		lenMapping := int(dataStartingAddrDiff)) + int(b.chipSize)
-		b.vPage, err := mmap.MapRegion(b.memFile, lenMapping, mmap.RDWR, 0, 0)
+		dataStartingAddrDiff := physAddr - pageStart
+		offset = int64(pageStart)
+		lenMapping = int(dataStartingAddrDiff) + int(chipSize)
+	}
 
+	/*
 		***** for the mmap() call ****
 		- if we were using /dev/mem, then offset = pageStart.
 		the file we 'open' starts at the base address of gpio memory for chip 0, not at the base of memory.
 		- we would access our memory by accessing vPage[dataStartingAddrDiff] if the start address of the data != page start address
-
 	*/
-
-	// 0 flag = shared, 0 offset because we are starting from the beginning of the mem/gpiomem0 file. offs = pageStart if we opened /dev/mem
-	b.vPage, err = mmap.MapRegion(b.memFile, int(b.chipSize), mmap.RDWR, 0, 0)
+	// 0 flag = shared
+	vPage, err := mmap.MapRegion(memFile, lenMapping, mmap.RDWR, 0, offset)
 	if err != nil {
-		return fmt.Errorf("failed to mmap: %w", err)
+		return Pinctrl{}, fmt.Errorf("failed to mmap: %w", err)
 	}
 
-	b.virtAddr = &b.vPage[0]
-	return err
+	return Pinctrl{MemFile: memFile, VPage: vPage, PhysAddr: physAddr, VirtAddr: &vPage[0]}, nil
 }
 
-// Sets up GPIO pin memory access by parsing the device tree for relevant address information.
-func (b *pinctrlpi5) setupPinControl(testingMode bool) error {
-	// TODO: "gpio0" is hardcoded as the gpioName.
+// SetupPinControl sets up GPIO pin memory access by parsing the device tree for relevant address information.
+// Implementers will need to contsruct a PinctrlConfig to use this.
+func SetupPinControl(cfg Config, logger logging.Logger) (Pinctrl, error) {
+	var err error
 	// This is not generalizeable; determine if there is a way to retrieve this from the pi / config / mapping information instead.
 
 	// If we are running tests, we need to read files/folders from our module's local sample device tree.
-	// This is located at raspi5-pinctrl/pi5/mock-device-tree
-	if testingMode {
-		dtBaseNodePath = testFolder + dtBase
-	} else {
-		dtBaseNodePath = dtBase
-	}
+	// This is located in mock-device-tree
+	testingMode := cfg.TestPath != ""
+	dtBaseNodePath := cfg.getBaseNodePath()
 
-	nodePath, err := b.findPathFromAlias(gpioName)
-	if err != nil {
-		b.logger.Errorf("error getting raspi5 GPIO nodePath")
-		return err
+	nodePath := cfg.GPIOChipPath
+	if cfg.UseAlias {
```

Comment:
its not necessary to do this or `setGPIONodePhysAddr` below if `UseGPIOMem` is set to true.